### PR TITLE
Export golang prometheus metrics in the cmdriver

### PR DESCRIPTION
**Pull request by marceloamaral**
The are other minor changes as well

### CODE REVIEW

File: cmdriver/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+FROM ubuntu:16.04
```
Comment (MikeSpreitzer) [inline cmdriver/Dockerfile]: Why so old?

File: cmdriver/main.go [hunk 2]
```diff
@@ -266,3 +275,9 @@ func writelog(op string, key string, tBefore, tAfter time.Time, csvFile *os.File
 	// fmt.Printf("%s,%s,%s,%q,%q\n", formatTime(tBefore), formatTime(tAfter), op, key, errS)
 	csvFile.Write([]byte(fmt.Sprintf("%s,%s,%s,%q,%q\n", formatTime(tBefore), formatTime(tAfter), op, key, errS)))
 }
+
+func exposeMetrics() {
+  http.Handle("/metrics", promhttp.Handler())
+  glog.Infof("starting HTTP server on port :%s\n", *metricPort)
```
Comment (MikeSpreitzer) [inline cmdriver/main.go]: These could be in the main goroutine.
That leaves no need to define a separate func here, just have the main goroutine say
```
    go glog.Fatal(http.ListenAndServe(":"+*metricPort, nil))
```

File: glide.yaml [hunk 3]
```diff
@@ -13,7 +13,6 @@ import:
   version: ^v9
   subpackages:
   - kubernetes
-  - rest
```
Comment (MikeSpreitzer) [inline glide.yaml]: Why remove this entry?  There still is a corresponding import.

File: glide.yaml [hunk 4]
```diff
@@ -26,3 +25,9 @@ import:
   version: ^0.9
   subpackages:
   - prometheus
+  - prometheus/promhttp
+- package: golang.org/x/sys
+  subpackages:
+  - unix
```
Comment (MikeSpreitzer) [inline glide.yaml]: Why add this to glide.yaml?  I do not see the go code importing it.

File: glide.yaml [hunk 5]
```diff
@@ -26,3 +25,9 @@ import:
...
+- package: github.com/prometheus/procfs
```
Comment (MikeSpreitzer) [inline glide.yaml]: Why add this to glide.yaml?  I do not see the go code importing it.

File: cmdriver/cmdriver.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
+apiVersion: v1
+kind: Service
```
Comment (MikeSpreitzer) [inline cmdriver/cmdriver.yaml]: Why is this here?  It suffices to annotate the Pod.  For example:
```
apiVersion: v1
kind: Pod
metadata:
  name: promtest
  annotations:
    prometheus.io/scrape: "true"
    prometheus.io/port: "9101"
spec:
  containers:
  - image: busybox
    name: bb
    command: [ "sleep", "86400" ]
```

**Review by MikeSpreitzer — COMMENTED** [review]
I am not a big fan of making a long-standing Pod that will be used for multiple runs.  I think it would be more natural for a Pod to do one run.  `kubectl run --restart=Never --image=mspreitz/cmdriver` will make a Pod, and `-l` can be used to apply labels.  We could expand the Prometheus configuration to look at labels as well as annotations, for Pods.  We could add a 15 second sleep at the end of the driver, to ensure the final state gets scraped.

**Comment by MikeSpreitzer** [comment]
Do you know what it would take to make a generator (see https://kubernetes.io/docs/reference/kubectl/conventions/#generators) specifically for this?

File: cmdriver/Dockerfile (hunk 1 above)
Comment (marceloamaral) [reply cmdriver/Dockerfile]: The other Dockerfiles use that version, I just wanna make it homogeneous.
But we can update all.

File: cmdriver/main.go (hunk 2 above)
Comment (marceloamaral) [reply cmdriver/main.go]: I agree

File: glide.yaml (hunk 3 above)
Comment (marceloamaral) [reply glide.yaml]: I did glide update and for some reason it was removed.
But I will add it back

File: glide.yaml (hunk 4 above)
Comment (marceloamaral) [reply glide.yaml]: Again, when I updated glide it was added, i will remove it

File: cmdriver/cmdriver.yaml (hunk 6 above)
Comment (marceloamaral) [reply cmdriver/cmdriver.yaml]: I tried before without the services and prometheus was not scraping the metrics....

File: cmdriver/main.go [hunk 7]
```diff
@@ -33,14 +36,16 @@ const (
...
+var kubeconfigPath = flag.String("kubeconfig", "", "Path to kubeconfig file, if empty the in cluster configuration will be used")
 var lambda = flag.Float64("lambda", 1.0, "Rate (1/s) at which new objects are created")
 var n = flag.Int("n", 300, "Total number of objects to create")
 var maxpop = flag.Int("maxpop", 100, "Maximum object population in system")
 var dataFilename = flag.String("datafile", "{{.RunID}}-driver.csv", "Name of CSV file to create")
 var runID = flag.String("runid", "", "unique ID of this run (default is randomly generated)")
 var seed = flag.Int64("seed", 0, "seed for random numbers (other than runid) (default is based on time)")
 var clientLB = flag.Bool("clientlb", false, "Load balance in this client")
+var metricPort = flag.String("metricport", "9101", "Port to expose prometheus metrics")
+var waitBeforeTerminate = flag.Int64("waitBeforeTerminate", 0, "Time in seconds to wait before terminate the program to have the metrics scraped by Prometheus")
```
Comment (MikeSpreitzer) [inline cmdriver/main.go:48]: I would make the default here 15 seconds.

File: cmdriver/main.go [hunk 8]
```diff
@@ -51,6 +56,11 @@ func main() {
 	flag.Set("logtostderr", "true")
 	flag.Parse()
 
+	// Start the HTTP server to expose golang prometheus metrics
+	http.Handle("/metrics", promhttp.Handler())
+	glog.Infof("starting HTTP server on port :%s\n", *metricPort)
+	go http.ListenAndServe(":"+*metricPort, nil)
```
Comment (MikeSpreitzer) [inline cmdriver/main.go:62]: Any error that is returned here should be reported.

File: cmdriver/cmdriver.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        - name: LOGDIR
+          value: /results
+        - name: WAITBEFORETERMINATE
+          value: "15"
+        ports:
+        - containerPort: 9101
+        volumeMounts:
+        - mountPath: /results
+          name: results
+      restartPolicy: Never
```
Comment (MikeSpreitzer) [inline cmdriver/cmdriver.yaml:43]: This means that if/when the driver crashes, the container will not be restarted within the same Pod but rather the Pod will be destroyed and a replacement created.  I do not think we want any form of retry.  Perhaps adding `backoffLimit: 0` would accomplish that.

File: cmdriver/cmdriver.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        command:
+          - "/cmdriver"
+          - "-lambda=$(LAMBDA)"
+          - "-maxpop=$(MAXPOP)"
+          - "-n=$(NUM)"
+          - "-log_dir=$(LOGDIR)"
+          - "-waitBeforeTerminate=$(WAITBEFORETERMINATE)"
+        env:
+        - name: LAMBDA
+          value: "10"
```
Comment (MikeSpreitzer) [inline cmdriver/cmdriver.yaml:29]: How does one use this with different values for these parameters?  Perhaps the Deployment was more convenient in this regard.

If we keep a single-run Job then it would perhaps be better to specify each environment variable on one line, like this:
```
    - {name: "LAMBDA", value: "10"}
```
That way there is some hope of automating edits to this.
```

**Comment by MikeSpreitzer** [comment]
I took most of this in #3 and merged it.  I find both the Deployment and the Job approaches problematic.  I think easiest will be to change the Prometheus configuration to scrape the bare process.

**Comment by MikeSpreitzer** [comment]
We went with #3 instead.
